### PR TITLE
Decimal_128 type checks

### PR DESCRIPTION
**Pull request by revans2**
This updates the type checks to be able to support 128-bit decimal. It mostly updates makes a distinction between decimal 64 support and decimal 128 support.  It also updates how we report both decimal and timestamp incompatibility in the docs and decimal incompatibility in the query explain.  The former is probably the biggest change and I would really like feedback on how it looks. This does cause a lot of merge conflicts so it might be best to hold off on this until 21.08 is really done.

### CODE REVIEW

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 1]
```diff
@@ -415,7 +505,12 @@ object TypeSig {
   val DATE: TypeSig = new TypeSig(TypeEnum.ValueSet(TypeEnum.DATE))
   val TIMESTAMP: TypeSig = new TypeSig(TypeEnum.ValueSet(TypeEnum.TIMESTAMP))
   val STRING: TypeSig = new TypeSig(TypeEnum.ValueSet(TypeEnum.STRING))
-  val DECIMAL: TypeSig = new TypeSig(TypeEnum.ValueSet(TypeEnum.DECIMAL))
+  val DECIMAL_64: TypeSig =
+    new TypeSig(TypeEnum.ValueSet(TypeEnum.DECIMAL), DType.DECIMAL64_MAX_PRECISION)
+  val DECIMAL_128_NO_OVERFLOW: TypeSig =
+    new TypeSig(TypeEnum.ValueSet(TypeEnum.DECIMAL), DecimalType.MAX_PRECISION - 1)
+  val DECIMAL_128_FULL: TypeSig =
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:512]: I'm a bit confused on the semantics for these, and I think a comment could help.  For example, is DECIMAL_128_NO_OVERFLOW covering cases where we need to check and prevent silent overflow or where we don't need to check?

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:512]: I thought about removing it entirely until we are ready to start using it. But then `DECIMAL_128_FULL` felt a little odd.  I will clean it up.

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Review by kuhushukla — APPROVED** [review]

**Comment by revans2** [comment]
build

**Review by jlowe — DISMISSED** [review]

**Review by jlowe — APPROVED** [review]

**Review by razajafri — DISMISSED** [review]

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
Hopefully this is the last time I am going to need to upmerge. If the build passes before something else gets merged in that causes a conflict I will merge this.

**Review by jlowe — APPROVED** [review]

**Pull request merged**